### PR TITLE
feat: add support for katapult security groups

### PR DESCRIPTION
**Pull request by andrewjtait**
This PR has the same changes as https://github.com/krystal/go-katapult/pull/102 but with the base branch using `references`. I've had a stab at making changes to match the changes introduced in `references`, but the tests are currently failing.

@strideynet - if you get a chance tomorrow maybe you could give me a hand in resolving the issues?

### CODE REVIEW

**Comment by andrewjtait** [comment]
@strideynet never mind, found the issue 😄

**Review by strideynet — CHANGES_REQUESTED** [review]
This looks mostly good to me, although @jimeh probably knows more about the actual Katapult API so his review here would be nice.

I think my only point is it would be nice if the tests were rewritten to use the new mock. If you see the LoadBalancerRule tests, we are no longer using the mux/handler technique and are now using a custom mocker we have built. However, if there's a rush for this/you are strife for time, I'm happy to accept the old style tests since they do work.

**Comment by jimeh** [comment]
I haven't had a chance to look at this yet, but just want to say if tests are already written, I think it's fine to leave them as is for now. #99 will be a refactor of the client interface and fakeRequestMaker, meaning tests using it will need to be refactored.

**Review by strideynet — APPROVED** [review]

File: core/security_group.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+type SecurityGroupUpdateRequest struct {
+	SecurityGroup SecurityGroupRef              `json:"security_group"`
+	Properties    *SecurityGroupUpdateArguments `json:"properties,omitempty"`
+}
+
+type SecurityGroupsResponseBody struct {
+	Pagination     *katapult.Pagination `json:"pagination,omitempty"`
+	SecurityGroup  *SecurityGroup       `json:"security_group,omitempty"`
+	SecurityGroups []*SecurityGroup     `json:"security_groups,omitempty"`
+}
```
Comment (jimeh) [inline core/security_group.go:63]: These three structs don't need to be exported, as they're purely used for internal over the wire data handling.

File: core/security_group.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+func (sgc *SecurityGroupsClient) GetByID(
+	ctx context.Context,
+	id string,
+) (*SecurityGroup, *katapult.Response, error) {
+	u := &url.URL{Path: fmt.Sprintf("security_groups/%s", id)}
+	body, resp, err := sgc.doRequest(ctx, "GET", u, nil)
+
+	return body.SecurityGroup, resp, err
+}
```
Comment (jimeh) [inline core/security_group.go:109]: To keep things more consistent, we're moving towards just always sending lookup refs via query strings. So I'd like to suggest the below changes, which will also require minor updates to the tests.

```suggestion
func (sgc *SecurityGroupsClient) Get(
	ctx context.Context,
	ref SecurityGroupRef,
) (*SecurityGroup, *katapult.Response, error) {
	u := &url.URL{
		Path:     "security_groups/_",
		RawQuery: ref.queryValues().Encode(),
	}
	body, resp, err := sgc.doRequest(ctx, "GET", u, nil)

	return body.SecurityGroup, resp, err
}

func (sgc *SecurityGroupsClient) GetByID(
	ctx context.Context,
	id string,
) (*SecurityGroup, *katapult.Response, error) {
	return sgc.Get(ctx, SecurityGroupRef{ID: id})
}
```

**Review by jimeh — CHANGES_REQUESTED** [review]

File: core/security_group.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+// allows a reference to a security group
+type SecurityGroupRef struct {
+	ID string `json:"id,omitempty"`
+}
+
+func (sg SecurityGroupRef) queryValues() *url.Values {
+	v := &url.Values{}
+	v.Set("security_group[id]", sg.ID)
+
+	return v
```
Comment (jimeh) [inline core/security_group.go:31]: Nitpick; As `url.Values` is just a `map[string][]string` type, and we're only ever setting one value, we can simplify this method like so:

```suggestion
	return &url.Values{
		"security_group[id]": []string{sg.ID},
	}
```

Personally I don't mind either way though, so I'll let you do with this suggestion as you wish :D

**Review by jimeh — COMMENTED** [review]
Changes look good :)... However, we should probably wait with merging this till #98 is merged, and also re-target it against the `main` branch at that point. Otherwise this feature will probably get lost in what I'm pretty sure will be a squash merge for #98.

Also, when this gets to a point of being merged, I'd like to suggest suggest using `security_groups` as the commit message scope, something like this:

```
feat(security_groups): add support for Security Groups
```

**Comment by andrewjtait** [comment]
@jimeh rebased and ready to roll ⚡

**Review by strideynet — APPROVED** [review]

File: core/security_group_test.go [hunk 4]
```diff
@@ -0,0 +1,1128 @@
...
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			testJSONMarshaling(t, tt.obj)
+		})
+	}
+}
+
+func TestSecurityGroup_queryValues(t *testing.T) {
```
Comment (jimeh) [inline core/security_group_test.go:64]:
```suggestion
func TestSecurityGroupRef_queryValues(t *testing.T) {
```

File: core/security_group_test.go [hunk 5]
```diff
@@ -0,0 +1,1128 @@
...
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			testJSONMarshaling(t, tt.obj)
+		})
+	}
+}
+
+func Test_SecurityGroupCreateRequest_JSONMarshaling(t *testing.T) {
```
Comment (jimeh) [inline core/security_group_test.go:167]:
```suggestion
func Test_securityGroupCreateRequest_JSONMarshaling(t *testing.T) {
```

File: core/security_group_test.go [hunk 6]
```diff
@@ -0,0 +1,1128 @@
...
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			testJSONMarshaling(t, tt.obj)
+		})
+	}
+}
+
+func Test_SecurityGroupUpdateRequest_JSONMarshaling(t *testing.T) {
```
Comment (jimeh) [inline core/security_group_test.go:191]:
```suggestion
func Test_securityGroupUpdateRequest_JSONMarshaling(t *testing.T) {
```

File: core/security_group_test.go [hunk 7]
```diff
@@ -0,0 +1,1128 @@
...
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			testJSONMarshaling(t, tt.obj)
+		})
+	}
+}
+
+func Test_SecurityGroupsResponseBody_JSONMarshaling(t *testing.T) {
```
Comment (jimeh) [inline core/security_group_test.go:215]:
```suggestion
func Test_securityGroupsResponseBody_JSONMarshaling(t *testing.T) {
```

File: core/security_group_test.go [hunk 8]
```diff
@@ -0,0 +1,1128 @@
...
+		{
+			name: "by organization ID",
+			args: args{
+				ctx: context.Background(),
+				org: OrganizationRef{ID: "org_O648YDMEYeLmqdmn"},
+			},
+			want: SecurityGroupList,
+			wantQuery: &url.Values{
+				"organization[id]": []string{"org_O648YDMEYeLmqdmn"},
+			},
```
Comment (jimeh) [inline core/security_group_test.go:285]: `wantQuery` is not actually used within the test body, as it does this instead to validate the request query string:

```go
qs := queryValues(tt.args.org, tt.args.opts)
assert.Equal(t, *qs, r.URL.Query())
```

We currently have a bit of a mixture of `wantQuery` and the above method between different tests. I'm undecided over which method I prefer, though I think I'm leaning a little bit more towards the `wantQuery` variant as it makes the test case more obvious about what's going on, and relies less on the various `queryValues()` methods working correctly.

@strideynet, @andrewjtait, thoughts?

Though is probably a discussion that's outside of scope of this PR, and simply removing `wantQuery` from this test is fine to get this merged :)

File: core/security_group_test.go [hunk 9]
```diff
@@ -0,0 +1,1128 @@
...
+		args       args
+		reqBody    *securityGroupCreateRequest
+		want       *SecurityGroup
+		errStr     string
+		errResp    *katapult.ResponseError
+		respStatus int
+		respBody   []byte
+	}{
+		{
+			name: "security group",
```
Comment (jimeh) [inline core/security_group_test.go:668]:
```suggestion
			name: "by organization ID",
```

File: core/security_group_test.go [hunk 10]
```diff
@@ -0,0 +1,1128 @@
...
+				Name:             "api-test",
+				AllowAllInbound:  true,
+				AllowAllOutbound: true,
+				Associations:     []string{"id1", "id2"},
+			},
+			respStatus: http.StatusCreated,
+			respBody:   fixture("security_group_create"),
+		},
+		{
+			name: "organization by sub-domain",
```
Comment (jimeh) [inline core/security_group_test.go:689]:
```suggestion
			name: "by organization SubDomain",
```

**Review by jimeh — CHANGES_REQUESTED** [review]
I've only got a few minor nitpicks about test function/case naming to keep things consistent, along with one comment about query string assertion that might need a bit of a discussion. Though said discussion will probably end up falling outside the scope of this PR.

File: core/client.go [hunk 11]
```diff
@@ -19,6 +19,7 @@ type Client struct {
 	NetworkSpeedProfiles            *NetworkSpeedProfilesClient
 	Networks                        *NetworksClient
 	Organizations                   *OrganizationsClient
+	SecurityGroups                  *SecurityGroupsClient
```
Comment (jimeh) [inline core/client.go:22]: A new `SecurityGroupsClient` instance needs to also be assigned to the `Client` isntance in the `New()` function.

File: core/security_group_test.go (hunk 8 above)
Comment (strideynet) [reply core/security_group_test.go:285]: I think I lean towards the explicit testing of the query values (this is what I've done in tests using the FRM)

File: core/security_group.go [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+func (sg *SecurityGroup) Ref() SecurityGroupRef {
+	return SecurityGroupRef{ID: sg.ID}
+}
+
+// allows a reference to a security group
+type SecurityGroupRef struct {
+	ID string `json:"id,omitempty"`
+}
```
Comment (jimeh) [inline core/security_group.go:25]: There's no JSON marshaling tests for this struct, though I've noticed that's actually an oversight missing from nearly every single `*Ref` type we have, and needs to be addressed. But let's add it for security groups here to begin with :)

**Review by jimeh — CHANGES_REQUESTED** [review]
This looks good, only thing left is some left-over golden files. That's easily fixed with:

```bash
make regen-golden
```

And then just commit the file deletions :)

**Comment by andrewjtait** [comment]
@jimeh this should be ready for re-review. I did push the commit yesterday morning but it appears to have got lost during the GH outage. I just toggled the base branch on the PR back and forth and now it has picked it up 🙃

**Review by jimeh — APPROVED** [review]

**Pull request merged**